Update WCPay welcome page incentives

ismaeldcom opened the pull request:
Closes https://github.com/Automattic/woocommerce-payments-server/issues/3479
Closes https://github.com/Automattic/woocommerce-payments-server/issues/3480

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:
As part of the new WCPay incentives projects (paJDYF-95I-p2 and paJDYF-96z-p2) we're going to expand the existing incentive to more countries and add another one. Reusing the existing implementation, but moving most of the logic to [WCPay server](https://github.com/Automattic/woocommerce-payments-server/issues/3498), allowing us to adjust the incentives without a need for additional WooCommerce PRs.

- Refactor the existing components a bit, to make them easier to maintain.
- Implement the new design (EJyQdHyExlXuiniHADVXEC-fi-39_7921).
- Compress the base64 menu icon.
- Fetch and cache eligible incentive from WCPay API.
  - Request from a public WPCOM endpoint with the following parameters:
    - `country` – Store ISO-2 country code, e.g. `US`.
    - `locale` – Store locale, e.g. `en_US`.
    - `active_for` – WooCommerce install timestamp.
    - `has_orders` – Whether the store has completed orders.
    - `has_payments` – Whether the store has at least one payment gateways enabled.
    - `has_wcpay` – Whether the store has a WooCommerce Payments account or it's installed.
  - Cache the response in a private variable, and `wcpay_welcome_page_incentive` transient by the given time or 24 hours for success requests, and 6 hours if there is an error.
- Filter `woocommerce_admin_shared_settings` to preload server-side incentive details on JS. Which are:
    - `id` – For allowed promo notes.
    - `description` – Incentive localized description copy. Allowing to render some HTML tags using `sanitizeHTML`.
    -  `cta_label` – Incentive localized CTA for the "install" button.
    -  `tos_url` – Terms of services URL.
- Add `woocommerce_admin_allowed_promo_notes` filter on `Notes`, to be able to easily add the server incentive ID on the protected `$allowed_promo_notes`.
- Add incentive ID to page view event, in order to differentiate every incentive in funnels.
- Update used option names and ensure they are added to `get_default_option_permissions`.
- Update dismiss logic to differentiate every incentive.
- Update delayed notes to use the logic from `WcPayWelcomePage` to check their visibility, to be able to check for server-side IDs.
- Fix a word case typo in `WCPayBannner`.

> **Note**: Incentives are still only shown when WooCommerce marketplace suggestions are enabled.

### How to test the changes in this Pull Request:

#### Setup
- To see the welcome incentive page, you need to meet (or comment [their code](https://github.com/woocommerce/woocommerce/blob/242ee8e921554185d32e4976518edd719f1a2bad/plugins/woocommerce/src/Internal/Admin/WcPayWelcomePage.php#L54)) the following requirements:
  - Have **WooCommerce → Settings → Advanced → WooCommerce.com → Display suggestions withing WooCommerced** enabled.
  - Do not have WCPay installed or an existing account cached.
  - Do not have it dismissed by `wcpay_welcome_page_incentives_dismissed` option.

#### Switch incentive
- Update WCPay server `Incentive_Service` to return switch incentive.
-  **Payments (1)** should be visible on the admin sidebar.
- Clicking on it should display the switch incentive offer, matching the attached screenshot.
- Choosing `Install WooPayments for free` should successfully create the promo note. Look for the incentive ID on `wc_admin_notes` table.

#### Cache
- Look for a transient with name `wcpay_welcome_page_incentive` on `options` table, and remove it if you want to force a fetch.
- Cache timestamp should match:
  - 6h for an WP Error.
  - Given time from server response or 24h.
  - No cache if the server returns `0`.

#### Action incentive
- Update WCPay server `Incentive_Service` to return action incentive.
-  **Payments (1)** should be visible on the admin sidebar.
- Clicking on it should display the switch incentive offer, matching the attached screenshot.
- Choosing `Install WooPayments for free` should successfully create the promo note. Look for incentive ID on `wc_admin_notes` table.

#### Updated Track
- Viewing the incentive page should record the `page_view` event with:
  - `path`: `payments_connect_core_test`
  - `incentive_id`: ID of the available incentive.
- You can easily check recorded Tracks by one of the following methods, ensuring `Allow usage of WooCommerce to be tracked` is checked:
  - Run `localStorage.setItem( 'debug', 'wc-admin:tracks' );` on your site browser console. And look for verbose logs.
  - Install [Tracks Vigilante](https://github.com/Automattic/tracks-chrome-extension) browser extension and use its log of events.
  - Go to [Tracks Live view](https://mc.a8c.com/tracks/live/), filter by any of the allowed methods, and wait patiently (~10min) for events to be registered there.

#### Dismiss survey and more info needed inbox note
- Click **No thanks**, then select **I need more information about WooCommerce Payments**, then click **Dismiss and send feedback**.
- Update `wcpay_welcome_page_exit_survey_more_info_needed_timestamp` option to a timestamp older than 30 days ago, `1` will work.
- Manually run the cron event `wc_admin_daily`. You can use a plugin like WP Crontrol.
- Go to **WooCommerce → Home** and verify that you see the new Inbox Note `Payments made simple with WooCommerce Payments` title.

#### Remind me later inbox note.
- Having an eligible incentive, and the **Payments** menu visible.
- Update `wcpay_welcome_page_viewed_timestamp` option to a timestamp older than 3 days ago, `1` will work.
- Manually run the cron event `wc_admin_daily`. You can use a plugin like WP Crontrol.
- Go to **WooCommerce → Home** and verify that you see the new Inbox Note `Save big with WooCommerce Payments` title.

#### Design
| Switch incentive | Action incentive |
|-|-|
|<img width="692" alt="Screenshot 2023-06-14 at 14 09 58" src="https://github.com/woocommerce/woocommerce/assets/7670276/13ba5104-844c-4ea2-958f-4638c0e25a55">|<img width="693" alt="Screenshot 2023-06-14 at 14 25 47" src="https://github.com/woocommerce/woocommerce/assets/7670276/b8dca3e1-a85d-4660-8937-db6aa9085283">|

ismaeldcom committed:
Initial refactor of components

Move out from index every UI component to make it easier to maintain.

ismaeldcom committed:
Initial implementation of new design

Some minor adjustments are missing, and I've added a few TODOs to complete later.

ismaeldcom committed:
Add changelog entry

ismaeldcom committed:
Finish design update

ismaeldcom committed:
Compress menu icon

ismaeldcom committed:
Add user's first name to heading copy

ismaeldcom committed:
Fetch and cache eligible incentive from WCPay API

Sharing its properties under `window.WcSettings.admin`

ismaeldcom committed:
Filter `allowed_promo_notes`

To be able to add those form WCPay API request

ismaeldcom committed:
Update JS side to use the new shared incentive details

ismaeldcom committed:
Ensure we cache requests even if there are no eligible incentives

ilyasfoo reviewed (commented):
@ismaeldcom It’s looking good so far! I really like the use of `cache-for` header to dynamically control cache server side. I think it's a pattern we should adopt more in core.

The only feedback I’d have is the use of `getAdminSetting` here. The ideal use case of settings would be something that is used globally. As I understood, the WCPay incentive is specifically used in the WCPay welcome page banner.

In my opinion, a better way to do it is via a new API endpoint which is called by the welcome page. This will limit the potential performance impact (albeit low) to only the relevant pages.

ismaeldcom committed:
Fix a typo in WCPayBanner

ismaeldcom committed:
Move incentive preloaded data from wcSettings

to a new window variable which will be only loaded on WCPay welcome page.

ismaeldcom committed:
Simplify `has_wcpay` check

ismaeldcom committed:
Render sanitized HTML from incentive description

ismaeldcom committed:
Update get incentive logic to also cache non success responses

ismaeldcom committed:
Add incentive ID to page view track

ismaeldcom committed:
Add changelog entry for the onboarding package

ismaeldcom committed:
Update used options names

and add them to `get_default_option_permissions

ismaeldcom committed:
Differentiate every dismissed incentive

Reuse WCPay welcome page logic for notes display conditions

ilyasfoo commented on plugins/woocommerce/src/Admin/API/Notes.php:
```diff
@@ -39,9 +39,22 @@ class Notes extends \WC_REST_CRUD_Controller {
+		 * @param array     $promo_notes    Array of allowed promo notes.
+		 * @since 7.8.0
+		 */
+		$this->allowed_promo_notes = apply_filters( 'woocommerce_admin_allowed_promo_notes', $this->allowed_promo_notes );
```
Do we have a reason to call this in a constructor? I think this is prone to a race condition, i.e.: a filter added after this class had already initialized. I'd suggest for calling the filter when the value is to be used.

ismaeldcom replied:
You're right, I was also worried about having a race condition here, although there are not, there is no reason to keep it here, as we only use `allowed_promo_notes` inside `activate_promo_note`. So moved it there in 8c26e80. Thanks for raising it.

ilyasfoo reviewed (commented):
Great work!

I wasn't able to get my local WCPay server running, I tested it with hardcoded values for the incentives instead. I've bumped into an issue with the incentive data not being available when we navigate to the welcome page via react router. 

To reproduce:

1. Go to WooCommerce > Home
2. Navigate to payments

```
react-dom.js?ver=18.2.0:22849 Uncaught TypeError: Cannot read properties of undefined (reading 'id')
    at index.tsx:69:28
    at commitHookEffectListMount (react-dom.js?ver=18.2.0:23160:28)
    at commitPassiveMountOnFiber (react-dom.js?ver=18.2.0:24936:15)
    at commitPassiveMountEffects_complete (react-dom.js?ver=18.2.0:24901:11)
    at commitPassiveMountEffects_begin (react-dom.js?ver=18.2.0:24888:9)
    at commitPassiveMountEffects (react-dom.js?ver=18.2.0:24876:5)
    at flushPassiveEffectsImpl (react-dom.js?ver=18.2.0:27049:5)
    at flushPassiveEffects (react-dom.js?ver=18.2.0:26994:16)
    at performSyncWorkOnRoot (react-dom.js?ver=18.2.0:26086:5)
    at flushSyncCallbacks (react-dom.js?ver=18.2.0:12052:24)
```

Apologies for potentially going back to a previous solution, but I think we'll need to have `wcpayWelcomePageIncentive` available globally wherever react app is present for us to do this without loading.

Another way is to load the data after react app is loaded, but I think it would not be as efficient.

vladolaru committed:
Use tos_url instead of tos_link to match response

vladolaru committed:
Fix incentives endpoint request payload

vladolaru commented:
Excellent work, @ismaeldcom!

I found a few inconsistencies that I think would be good to be addressed:
1. On the "No, thanks" exit modal we should use "WooPayments" instead of WooCommerce Payments.
2. Also on that modal, there is a padding-bottom of 100px. Is that intentional?
3. When you dismiss the offer the menu item is still available and and it leads to a blank page. On a second refresh it is gone.

ismaeldcom committed:
Back to preload incentive on admin shared settings

ismaeldcom committed:
Move promo notes filter to where it's used

to avoid any possible race conditions.

ismaeldcom commented:
Thanks for the quick review @ilyasfoo!

>Apologies for potentially going back to a previous solution, but I think we'll need to have wcpayWelcomePageIncentive available globally wherever react app is present for us to do this without loading.

Nice catch! And no problem, I didn't think about it either 🙂 __Addressed in 78799bf__

ismaeldcom committed:
Replace every occurrence of `WooCommerce Payments` with `WooPayments`

ismaeldcom commented:
Thanks for the review and commits @vladolaru!

>On the "No, thanks" exit modal we should use "WooPayments" instead of WooCommerce Payments.

Nice catch, I forgot to also apply the new branding there. _Addressed in [c177deb](https://github.com/woocommerce/woocommerce/pull/38689/commits/c177deb98640c7e2530c22b8be291c7b74123616)_

>Also on that modal, there is a padding-bottom of 100px. Is that intentional?
>When you dismiss the offer the menu item is still available and and it leads to a blank page. On a second refresh it is gone.

I'm just looking at it, it doesn't match what I see on local.

ismaeldcom committed:
Prevent modal large padding-bottom

ismaeldcom committed:
Merge branch 'trunk' into update/wcpay-welcome-page-incentives

ismaeldcom committed:
Prevent another global style override on modal

ismaeldcom committed:
Fix TS error coming from trunk merge

ismaeldcom committed:
Fix wrong style property

vladolaru committed:
Further restrict WCPay incentives to plugin not active and countries supported

vladolaru committed:
Adjust the meaning of has_wcpay

ismaeldcom committed:
Wait for dismiss option update before redirect

to prevent the side menu entry to be visible after redirecting back to WC admin.

vladolaru committed:
Enforce TOS asterix at the end of incentive description

vladolaru reviewed (approved):
With the latest changes that we've made, all things look great! I am approving this. Excellent work, Ismael!

ismaeldcom committed:
Remove underscore from priavate instance var.

ismaeldcom committed:
Fix ESLint error

ismaeldcom committed:
Fix SCSS linting issues

ismaeldcom committed:
Refine `has_wcpay` and `has_orders`

vladolaru committed:
Enforce consistent WooPayments branding

vladolaru committed:
Enforce stricter i18n calls with escape

ilyasfoo commented on plugins/woocommerce-admin/client/payments-welcome/index.tsx:
```diff
@@ -53,87 +42,44 @@ interface activatePromoResponse {
+			incentive_id: incentive.id,
+		} );
+		updateOptions( {
+			wcpay_welcome_page_viewed_timestamp: Math.floor(
```
The value is reset every time a user views this page, is that intended?

ismaeldcom replied:
Yes, it was only updated for the first time before, but it doesn't work as intended. Because we actually want to know that was the last time they viewed it, to send a follow-up notice 🙁

ismaeldcom committed:
Update page title to match new branding

ilyasfoo commented on plugins/woocommerce/src/Internal/Admin/WcPayWelcomePage.php:
```diff
@@ -4,90 +4,57 @@
  */
 class WcPayWelcomePage {
+	const TRANSIENT_NAME = 'wcpay_welcome_page_incentive';
+	const SCRIPT_NAME    = 'wcpay-welcome-page-incentive';
```
I can't find any usage reference for this, do we need it?

ismaeldcom replied:
Great catch, I didn't remove it after going back to shared settings. Addressed in 36a69ff

ilyasfoo commented on plugins/woocommerce/src/Internal/Admin/WcPayWelcomePage.php:
```diff
@@ -4,90 +4,57 @@
+
+		return self::$instance;
+	}
+	/**
```
Super nitpick 🙏 

```suggestion

	/**
```

vladolaru committed:
docs: Enforce branding in code docs also

vladolaru committed:
Remove method that is unreliable to be used outside of onboarding

ilyasfoo commented on plugins/woocommerce/src/Internal/Admin/WcPayWelcomePage.php:
```diff
@@ -4,90 +4,57 @@
-		}
-
-		// Suggestions may be disabled via a setting.
+	public function must_be_visible() {
```
Should we check if a user has already accepted an incentive for `welcome_page`? I imagine it would technically be possible that a user can accept multiple incentives otherwise.

ismaeldcom replied:
Actually, as the acceptance of any welcome page incentive will end with WCPay installed and an account created, it will not meet the conditions for the next incentive.

We also check on the server if by any chance the merchant try to apply a new welcome incentive while having another one 🙂

ilyasfoo reviewed (approved):
Tested well! I have some questions and nitpick comment, but this LGTM already 🚢

ismaeldcom committed:
Remove unused constant

ismaeldcom committed:
Missing new line

Co-authored-by: Ilyas Foo <foo.ilyas@gmail.com>
